src/utils/previewOutput.ts

```diff
@@ -0,0 +1,262 @@
+    .map((sc) => ({ fullName: sc.fullName, type: sc.type.name, path: ensureAbsolutePath(sc.xml) }));
+  // dedupe by xml path
+  return Array.from(new Map(sourceComponents.map((sc) => [sc.path, sc])).values());
+};
+
+const calculateDeployOperation = (destructiveChangesType?: DestructiveChangesType): PreviewFile['operation'] => {
+  switch (destructiveChangesType) {
+    case DestructiveChangesType.POST:
+      return 'deletePost';
+    case DestructiveChangesType.PRE:
+      return 'deletePre';
+    default:
+      return 'deploy';
+  }
+};
+
+const getNonIgnoredConflicts = (files: PreviewFile[]): PreviewFile[] => files.filter((f) => f.conflict && !f.ignored);
+
+const willGo = (previewFile: PreviewFile): boolean => !previewFile.conflict && !previewFile.ignored;
+
+const getWillDeploy = (files: PreviewFile[]): PreviewFile[] =>
+  files.filter((f) => willGo(f) && f.operation === 'deploy');
+
+const getWillRetrieve = (files: PreviewFile[]): PreviewFile[] =>
+  files.filter((f) => willGo(f) && f.operation === 'retrieve');
+
+const getWillDelete = (files: PreviewFile[]): PreviewFile[] =>
+  files.filter((f) => willGo(f) && ['deletePre', 'deletePost'].includes(f.operation));
+
+// relative paths are easier on tables
+const columns = { type: {}, fullName: {}, projectRelativePath: { header: 'Path' } };
+const makeKey = ({ type, fullName }: { type: MetadataType; fullName: string }): string => `${type.name}#${fullName}`;
+
+export const compileResults = ({
+  componentSet,
+  projectPath,
+  filesWithConflicts,
+  forceIgnore,
+  baseOperation,
+  remoteDeletes,
+}: {
+  componentSet: ComponentSet;
+  projectPath: string;
+  filesWithConflicts: Set<string>;
+  forceIgnore: ForceIgnore;
+  baseOperation: BaseOperation;
+  remoteDeletes?: SourceComponent[];
+}): PreviewResult => {
+  // when we iterate all the componentSet,
+  // this map makes it easy to get the source-backed local components
+  const sourceBackedComponents = new Map<string, SourceComponent>(
+    componentSet.getSourceComponents().map((sc) => [makeKey({ type: sc.type, fullName: sc.fullName }), sc])
+  );
+
+  const sourceComponentToPreviewFile = (c: SourceComponent): Omit<PreviewFile, 'operation'> => ({
+    type: c.type.name,
+    fullName: c.fullName,
+    conflict: [c.xml, c.content].some((v) => v && filesWithConflicts.has(v)),
+    ignored: [c.xml, c.content].some((v) => v && forceIgnore.denies(v)),
+    path: path.isAbsolute(c.xml) ? c.xml : path.resolve(c.xml), // SDR/SourceComponent uses absolute path
+    projectRelativePath: path.relative(projectPath, c.xml), // for cleaner output
+    // There should not be anything in forceignore returned by the componentSet
+  });
+
+  const actionableFiles = componentSet
+    .toArray()
+    .map((c): SourceComponent | MetadataComponent => sourceBackedComponents.get(makeKey(c)) ?? c)
+    .map(
+      (c): PreviewFile =>
+        'xml' in c
+          ? // source backed components exist locally
```

Comment:
I find ternaries like this difficult to read. My general rule is that if I can't fit on one line, I use `if/else`

Reply:
I combined these two comment into a branch.  It looks a lot worse to me, but I can live with it if you've got a strong preference about multi-line ternaries